src/boot/efi/initrd.c

```diff
@@ -0,0 +1,135 @@
+        .vendor = {
+                {
+                        MEDIA_DEVICE_PATH,
+                        MEDIA_VENDOR_DP,
+                        { sizeof(efi_initrd_device_path.vendor), 0 }
+                },
+                LINUX_INITRD_MEDIA_GUID
+        },
+        .end = {
+                END_DEVICE_PATH_TYPE,
+                END_ENTIRE_DEVICE_PATH_SUBTYPE,
+                { sizeof(efi_initrd_device_path.end), 0 }
+        }
+};
+
+EFIAPI EFI_STATUS initrd_load_file(
+                EFI_LOAD_FILE_PROTOCOL *this,
+                EFI_DEVICE_PATH *file_path,
+                BOOLEAN boot_policy,
+                UINTN *buffer_size,
+                VOID *buffer) {
+
+        if (!this || !buffer_size || !file_path)
+                return EFI_INVALID_PARAMETER;
+        if (boot_policy)
+                return EFI_UNSUPPORTED;
+
+        struct initrd_loader *loader = (struct initrd_loader *) this;
+
+        if (loader->length == 0 || !loader->address)
+                return EFI_NOT_FOUND;
+
+        if (!buffer || *buffer_size < loader->length) {
+                *buffer_size = loader->length;
+                return EFI_BUFFER_TOO_SMALL;
+        }
+
+        CopyMem(buffer, loader->address, loader->length);
+        *buffer_size = loader->length;
+        return EFI_SUCCESS;
+}
+
+EFI_STATUS initrd_register(
+                VOID *initrd_address,
+                UINTN initrd_length,
+                EFI_HANDLE *initrd_handle) {
+
+        EFI_STATUS err;
+        EFI_DEVICE_PATH *dp = (EFI_DEVICE_PATH *) &efi_initrd_device_path;
+        EFI_HANDLE handle;
+        struct initrd_loader *loader;
+
+        assert(initrd_handle);
+
+        if (!initrd_address || initrd_length == 0)
+                return EFI_SUCCESS;
+
+        /* check if a InitrdMedia DevicePath is alreay registed.
+           LocateDevicePath checks for the "closest DevicePath" and returns its handle,
+           where as InstallMultipleProtocolInterfaces only maches identical DevicePaths.
+         */
+        err = uefi_call_wrapper(BS->LocateDevicePath, 3, &EfiLoadFile2Protocol , &dp, &handle);
+        if (err != EFI_NOT_FOUND) /* InitrdMedia is already registered */
+                return EFI_ALREADY_STARTED;
+
+        loader = AllocatePool(sizeof(struct initrd_loader));
+        if (!loader)
+                return EFI_LOAD_ERROR;
+
+        loader->load_file.LoadFile = initrd_load_file;
+        loader->address = initrd_address;
+        loader->length = initrd_length;
+
+        /* create a new handle and register the LoadFile2 protocol with the InitrdMediaPath on it */
+        err = uefi_call_wrapper(BS->InstallMultipleProtocolInterfaces, 8,
+                                initrd_handle,
+                                &DevicePathProtocol, &efi_initrd_device_path,
+                                &EfiLoadFile2Protocol, loader,
+                                NULL);
+        if (EFI_ERROR(err)) {
+                *initrd_handle = NULL;
```

Comment:
so in the systemd codebase we generally try to follow the rule to not touch return parameters on failure, and always initialize them on success. Some underlying APIs don#t follow this rule and also touch return parameters in some form of failures. We usually try to hide that away, to make our own interfaces somewhat uniform on this. Usually this means that we operate on local, temporary variables for things, and only copy stuff into our own return parameters when nothing can fail anymore.